File: Software/src/battery/RANGE-ROVER-PHEV-BATTERY.cpp
#include "RANGE-ROVER-PHEV-BATTERY.h"
#include "../communication/can/comm_can.h"
#include "../datalayer/datalayer.h"
#include "../devboard/utils/events.h"
#include "../include.h"

/* TODO
- LOG files from vehicle needed to determine CAN content needed to send towards battery!
  - BCCM_PMZ_A (0x18B 50ms)
  - BCCMB_PMZ_A (0x224 90ms)
  - BCM_CCP_RX_PMZCAN (0x601 non cyclic)
  - EPIC_PMZ_B (0x009 non cyclic)
  - GWM_FuelPumpEnableDataControl_PMZ (0x1F8 non cyclic)
  - GWM_IgnitionAuthDataTarget_PMZ (0x004 non cyclic)
  - GWM_PMZ_A (0x008 10ms cyclic)
  - GWM_PMZ_B -F, G-I, Immo, K-P 
    - 0x010 10ms
    - 0x090 10ms
    - 0x108 20ms
    - 0x110 20ms
    - 0x1d0 80ms
    - 0x490 900ms
    - 0x1B0 80ms
    - 0x460 720ms
    - 0x006 non cyclic immo
    - 0x450 600ms
    - 0x2b8 180ms
    - 0x388 200ms
    - 0x2b0 180ms
    - 0x380 80ms
  - GWM_PMZ_V_HYBRID (0x18d 60ms)
  - HVAC_PMZ_A-E
    - 0x1a8 70ms
    - 0x210 100ms
    - 0x300 200ms
    - 0x440 180ms
    - 0x0c0 10ms 
  - PCM_PMZ_C_Hybrid C, D, H, M
    - 0x030 15ms
    - 0x304 180ms
    - 0x1C0 80ms
    - 0x434 350ms
  - TCU_PMZ_A
    - 0x014 non cyclic, command from TCU, most likely not needed
- Determine CRC calculation
- Figure out contactor closing requirements
*/

void RangeRoverPhevBattery::update_values() {

  datalayer.battery.status.real_soc = SOCAverage;

  datalayer.battery.status.soh_pptt = StateofHealth * 10;

  datalayer.battery.status.voltage_dV = VoltageExt * 10;

  datalayer.battery.status.current_dA = (CurrentExt * 0.025) - 209715;

  datalayer.battery.status.max_charge_power_W = (ChargeContPwrLmt * 10) - 6550;

  datalayer.battery.status.max_discharge_power_W = (DischargeContPwrLmt * 10) - 6550;

  datalayer.battery.status.remaining_capacity_Wh = static_cast<uint32_t>(
      (static_cast<double>(datalayer.battery.status.real_soc) / 10000) * datalayer.battery.info.total_capacity_Wh);

  datalayer.battery.status.cell_max_voltage_mV = CellVoltageMax;

  datalayer.battery.status.cell_min_voltage_mV = CellVoltageMin;

  datalayer.battery.status.temperature_min_dC = CellTempColdest * 10;

  datalayer.battery.status.temperature_max_dC = CellTempHottest * 10;

  datalayer.battery.info.max_design_voltage_dV = ChargeVoltageLimit * 10;

  datalayer.battery.info.min_design_voltage_dV = DischargeVoltageLimit * 10;
}

void RangeRoverPhevBattery::handle_incoming_can_frame(CAN_frame rx_frame) {
  datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
  switch (rx_frame.ID) {
    case 0x080:  // 15ms
      StatusCAT5BPOChg = (rx_frame.data.u8[0] & 0x01);
      StatusCAT4Derate = (rx_frame.data.u8[0] & 0x02) >> 1;
      OCMonitorStatus = (rx_frame.data.u8[0] & 0x0C) >> 2;
      StatusCAT3 = (rx_frame.data.u8[0] & 0x10) >> 4;
      IsolationStatus = (rx_frame.data.u8[0] & 0x20) >> 5;
      HVILStatus = (rx_frame.data.u8[0] & 0x40) >> 6;
      ContactorStatus = (rx_frame.data.u8[0] & 0x80) >> 7;
      StatusGpCounter = (rx_frame.data.u8[1] & 0x0F);
      WeldCheckStatus = (rx_frame.data.u8[1] & 0x20) >> 5;
      StatusCAT7NowBPO = (rx_frame.data.u8[1] & 0x40) >> 6;
      StatusCAT6DlyBPO = (rx_frame.data.u8[1] & 0x80) >> 7;
      StatusGpCS = rx_frame.data.u8[2];
      CAT6Count = rx_frame.data.u8[3] & 0x7F;
      EndOfCharge = (rx_frame.data.u8[6] & 0x04) >> 2;
      DerateWarning = (rx_frame.data.u8[6] & 0x08) >> 3;
      PrechargeAllowed = (rx_frame.data.u8[6] & 0x10) >> 4;
      break;
    case 0x100:  // 20ms
      DischargeExtGpCounter = (rx_frame.data.u8[0] & 0x0F);
      DischargeExtGpCS = rx_frame.data.u8[1];
      DischargeVoltageLimit = (((rx_frame.data.u8[2] & 0x03) << 8) | rx_frame.data.u8[3]);
      DischargePowerLimitExt = ((rx_frame.data.u8[4] << 8) | rx_frame.data.u8[5]);
      DischargeContPwrLmt = ((rx_frame.data.u8[6] << 8) | rx_frame.data.u8[7]);
      break;
    case 0x102:  // 20ms
      PwrGpCS = rx_frame.data.u8[0];
      PwrGpCounter = (rx_frame.data.u8[1] & 0x3C) >> 2;
      VoltageExt = (((rx_frame.data.u8[1] & 0x03) << 8) | rx_frame.data.u8[2]);
      VoltageBus = (((rx_frame.data.u8[3] & 0x03) << 8) | rx_frame.data.u8[4]);
      CurrentExt = ((rx_frame.data.u8[5] << 8) | (rx_frame.data.u8[6] << 8) | rx_frame.data.u8[7]);
      break;
    case 0x104:  // 20ms
      HVIsolationTestRunning = (rx_frame.data.u8[2] & 0x10) >> 4;
      VoltageOC = (((rx_frame.data.u8[2] & 0x03) << 8) | rx_frame.data.u8[3]);
      DchCurrentLimit = (((rx_frame.data.u8[4] & 0x03) << 8) | rx_frame.data.u8[5]);
      ChgCurrentLimit = (((rx_frame.data.u8[6] & 0x03) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x10A:  // 20ms
      ChargeContPwrLmt = ((rx_frame.data.u8[0] << 8) | rx_frame.data.u8[1]);
      ChargePowerLimitExt = ((rx_frame.data.u8[2] << 8) | rx_frame.data.u8[3]);
      ChgExtGpCS = rx_frame.data.u8[4];
      ChgExtGpCounter = (rx_frame.data.u8[5] >> 4);
      ChargeVoltageLimit = (((rx_frame.data.u8[6] & 0x03) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x198:  // 60ms
      CurrentWarning = (rx_frame.data.u8[4] & 0x03);
      TempWarning = ((rx_frame.data.u8[4] & 0x0C) >> 2);
      TempUpLimit = (rx_frame.data.u8[5] / 2) - 40;
      CellVoltWarning = ((rx_frame.data.u8[6] & 0x60) >> 5);
      CCCVChargeMode = ((rx_frame.data.u8[6] & 0x80) >> 7);
      CellVoltUpLimit = (((rx_frame.data.u8[6] & 0x1F) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x220:  // 100ms
      SOCHighestCell = (((rx_frame.data.u8[0] & 0x3F) << 8) | rx_frame.data.u8[1]);
      SOCLowestCell = (((rx_frame.data.u8[2] & 0x3F) << 8) | rx_frame.data.u8[3]);
      SOCAverage = (((rx_frame.data.u8[4] & 0x3F) << 8) | rx_frame.data.u8[5]);
      WakeUpTopUpReq = ((rx_frame.data.u8[6] & 0x04) >> 2);
      WakeUpThermalReq = ((rx_frame.data.u8[6] & 0x08) >> 3);
      WakeUpDchReq = ((rx_frame.data.u8[6] & 0x10) >> 4);
      StateofHealth = (((rx_frame.data.u8[6] & 0x03) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x308:  // 190ms
      EstimatedLossChg = (((rx_frame.data.u8[0] & 0x03) << 8) | rx_frame.data.u8[1]);
      CoolingRequest = ((rx_frame.data.u8[6] & 0x04) >> 2);
      EstimatedLossDch = (((rx_frame.data.u8[2] & 0x03) << 8) | rx_frame.data.u8[3]);
      FanDutyRequest = (rx_frame.data.u8[4] & 0x7F);
      ValveCtrlStat = ((rx_frame.data.u8[4] & 0x80) >> 7);
      EstLossDchTgtSoC = (((rx_frame.data.u8[5] & 0x03) << 8) | rx_frame.data.u8[6]);
      break;
    case 0x424:  // 280ms
      HeatPowerGenChg = (rx_frame.data.u8[0] & 0x7F);
      HeatPowerGenDch = (rx_frame.data.u8[1] & 0x7F);
      WarmupRateChg = (rx_frame.data.u8[2] & 0x3F);
      WarmupRateDch = (rx_frame.data.u8[3] & 0x3F);
      CellVoltageMax = (((rx_frame.data.u8[4] & 0x1F) << 8) | rx_frame.data.u8[5]);
      CellVoltageMin = (((rx_frame.data.u8[6] & 0x1F) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x448:  // 600ms
      CellTempAverage = (rx_frame.data.u8[0] / 2) - 40;
      CellTempColdest = (rx_frame.data.u8[1] / 2) - 40;
      CellTempHottest = (rx_frame.data.u8[2] / 2) - 40;
      HeaterCtrlStat = (rx_frame.data.u8[3] & 0x7F);
      ThermalOvercheck = ((rx_frame.data.u8[3] & 0x80) >> 7);
      InletCoolantTemp = rx_frame.data.u8[5];
      ClntPumpDiagStat_UB = ((rx_frame.data.u8[6] & 0x04) >> 2);
      InletCoolantTemp_UB = ((rx_frame.data.u8[6] & 0x08) >> 3);
      CoolantLevel = ((rx_frame.data.u8[6] & 0x10) >> 4);
      ClntPumpDiagStat = ((rx_frame.data.u8[6] & 0x20) >> 5);
      MILRequest = ((rx_frame.data.u8[6] & 0xC0) >> 6);
      break;
    case 0x464:  // 800ms
      EnergyAvailable = (((rx_frame.data.u8[0] & 0x07) << 8) | rx_frame.data.u8[1]);
      EnergyUsableMax = (((rx_frame.data.u8[2] & 0x07) << 8) | rx_frame.data.u8[3]);
      EnergyUsableMin = (((rx_frame.data.u8[4] & 0x07) << 8) | rx_frame.data.u8[5]);
      TotalCapacity = (((rx_frame.data.u8[6] & 0x0F) << 8) | rx_frame.data.u8[7]);
      break;
    case 0x5A2:  //Not periodically transferred
      break;
    case 0x656:  //Not periodically transferred
      break;
    case 0x657:  //Not periodically transferred
      break;
    case 0x6C8:  //Not periodically transferred
      break;
    case 0x6C9:  //Not periodically transferred
      break;
    case 0x6CA:  //Not periodically transferred
      break;
    case 0x6CB:  //Not periodically transferred
      break;
    case 0x7EC:  //Not periodically transferred
      break;
    default:
      break;
  }
}

void RangeRoverPhevBattery::transmit_can(unsigned long currentMillis) {
  // Send 50ms CAN Message
  if (currentMillis - previousMillis50ms >= INTERVAL_50_MS) {
    previousMillis50ms = currentMillis;

    transmit_can_frame(&RANGE_ROVER_18B, can_config.battery);
  }
}

void RangeRoverPhevBattery::setup(void) {  // Performs one time setup at startup
  strncpy(datalayer.system.info.battery_protocol, Name, 63);
  datalayer.system.info.battery_protocol[63] = '\0';
  datalayer.system.status.battery_allows_contactor_closing = true;
  datalayer.battery.info.max_design_voltage_dV = MAX_PACK_VOLTAGE_DV;
  datalayer.battery.info.min_design_voltage_dV = MIN_PACK_VOLTAGE_DV;
  datalayer.battery.info.max_cell_voltage_mV = MAX_CELL_VOLTAGE_MV;
  datalayer.battery.info.min_cell_voltage_mV = MIN_CELL_VOLTAGE_MV;
}

File: frontend/src/ext/datalayer/RangeRoverPhevBattery.ts
// RangeRoverPhevBattery: 216 bytes; base classes: CanBattery@0
export const RANGE_ROVER_PHEV_BATTERY_FIELDS: ([string, string] | [string, string, number])[] = [
  ['___vptr$Battery', '__vtbl_ptr_type*'],
  ['__defaultRenderer', 'BatteryDefaultRenderer'],
  ['', ' ', 4],
  ['___vptr$Transmitter', '__vtbl_ptr_type*'],
  ['___vptr$CanReceiver', '__vtbl_ptr_type*'],
  ['_can_interface', 'CAN_Interface'],
  ['_initial_speed', 'CAN_Speed'],
  ['previousMillis50ms', 'u32'],
  ['StatusCAT5BPOChg', 'b'],
  ['StatusCAT4Derate', 'b'],
  ['OCMonitorStatus', 'u8'],
  ['StatusCAT3', 'b'],
  ['IsolationStatus', 'b'],
  ['HVILStatus', 'b'],
  ['ContactorStatus', 'b'],
  ['StatusGpCounter', 'u8'],
  ['WeldCheckStatus', 'b'],
  ['StatusCAT7NowBPO', 'b'],
  ['StatusCAT6DlyBPO', 'b'],
  ['StatusGpCS', 'u8'],
  ['CAT6Count', 'u8'],
  ['EndOfCharge', 'b'],
  ['DerateWarning', 'b'],
  ['PrechargeAllowed', 'b'],
  ['DischargeExtGpCounter', 'u8'],
  ['DischargeExtGpCS', 'u8'],
  ['DischargeVoltageLimit', 'u16'],
  ['DischargePowerLimitExt', 'u16'],
  ['DischargeContPwrLmt', 'u16'],
  ['PwrGpCS', 'u8'],
  ['PwrGpCounter', 'u8'],
  ['VoltageExt', 'u16'],
  ['VoltageBus', 'u16'],
  ['', ' ', 2],
  ['CurrentExt', 'i32'],
  ['HVIsolationTestRunning', 'b'],
  ['', ' '],
  ['VoltageOC', 'u16'],
  ['DchCurrentLimit', 'u16'],
  ['ChgCurrentLimit', 'u16'],
  ['ChargeContPwrLmt', 'u16'],
  ['ChargePowerLimitExt', 'u16'],
  ['ChgExtGpCS', 'u8'],
  ['ChgExtGpCounter', 'u8'],
  ['ChargeVoltageLimit', 'u16'],
  ['CurrentWarning', 'u8'],
  ['TempWarning', 'u8'],
  ['TempUpLimit', 'i8'],
  ['CellVoltWarning', 'u8'],
  ['CCCVChargeMode', 'b'],
  ['', ' '],
  ['CellVoltUpLimit', 'u16'],
  ['SOCHighestCell', 'u16'],
  ['SOCLowestCell', 'u16'],
  ['SOCAverage', 'u16'],
  ['WakeUpTopUpReq', 'b'],
  ['WakeUpThermalReq', 'b'],
  ['WakeUpDchReq', 'b'],
  ['', ' '],
  ['StateofHealth', 'u16'],
  ['EstimatedLossChg', 'u16'],
  ['CoolingRequest', 'b'],
  ['', ' '],
  ['EstimatedLossDch', 'u16'],
  ['FanDutyRequest', 'u8'],
  ['ValveCtrlStat', 'b'],
  ['EstLossDchTgtSoC', 'u16'],
  ['HeatPowerGenChg', 'u8'],
  ['HeatPowerGenDch', 'u8'],
  ['WarmupRateChg', 'u8'],
  ['WarmupRateDch', 'u8'],
  ['CellVoltageMax', 'u16'],
  ['CellVoltageMin', 'u16'],
  ['CellTempAverage', 'i8'],
  ['CellTempColdest', 'i8'],
  ['CellTempHottest', 'i8'],
  ['HeaterCtrlStat', 'u8'],
  ['ThermalOvercheck', 'b'],
  ['InletCoolantTemp', 'i8'],
  ['ClntPumpDiagStat_UB', 'b'],
  ['InletCoolantTemp_UB', 'b'],
  ['CoolantLevel', 'b'],
  ['ClntPumpDiagStat', 'b'],
  ['MILRequest', 'u8'],
  ['', ' '],
  ['EnergyAvailable', 'u16'],
  ['EnergyUsableMax', 'u16'],
  ['EnergyUsableMin', 'u16'],
  ['TotalCapacity', 'u16'],
  ['', ' ', 6],
  ['RANGE_ROVER_18B', 'CAN_frame'],
];

export const ___vptr$Battery: [number, string] = [0, '__vtbl_ptr_type*'];
export const __defaultRenderer: [number, string] = [4, 'BatteryDefaultRenderer'];
export const ___vptr$Transmitter: [number, string] = [8, '__vtbl_ptr_type*'];
export const ___vptr$CanReceiver: [number, string] = [12, '__vtbl_ptr_type*'];
export const _can_interface: [number, string] = [16, 'CAN_Interface'];
export const _initial_speed: [number, string] = [20, 'CAN_Speed'];
export const previousMillis50ms: [number, string] = [24, 'u32'];
export const StatusCAT5BPOChg: [number, string] = [28, 'b'];
export const StatusCAT4Derate: [number, string] = [29, 'b'];
export const OCMonitorStatus: [number, string] = [30, 'u8'];
export const StatusCAT3: [number, string] = [31, 'b'];
export const IsolationStatus: [number, string] = [32, 'b'];
export const HVILStatus: [number, string] = [33, 'b'];
export const ContactorStatus: [number, string] = [34, 'b'];
export const StatusGpCounter: [number, string] = [35, 'u8'];
export const WeldCheckStatus: [number, string] = [36, 'b'];
export const StatusCAT7NowBPO: [number, string] = [37, 'b'];
export const StatusCAT6DlyBPO: [number, string] = [38, 'b'];
export const StatusGpCS: [number, string] = [39, 'u8'];
export const CAT6Count: [number, string] = [40, 'u8'];
export const EndOfCharge: [number, string] = [41, 'b'];
export const DerateWarning: [number, string] = [42, 'b'];
export const PrechargeAllowed: [number, string] = [43, 'b'];
export const DischargeExtGpCounter: [number, string] = [44, 'u8'];
export const DischargeExtGpCS: [number, string] = [45, 'u8'];
export const DischargeVoltageLimit: [number, string] = [46, 'u16'];
export const DischargePowerLimitExt: [number, string] = [48, 'u16'];
export const DischargeContPwrLmt: [number, string] = [50, 'u16'];
export const PwrGpCS: [number, string] = [52, 'u8'];
export const PwrGpCounter: [number, string] = [53, 'u8'];
export const VoltageExt: [number, string] = [54, 'u16'];
export const VoltageBus: [number, string] = [56, 'u16'];
export const CurrentExt: [number, string] = [60, 'i32'];
export const HVIsolationTestRunning: [number, string] = [64, 'b'];
export const VoltageOC: [number, string] = [66, 'u16'];
export const DchCurrentLimit: [number, string] = [68, 'u16'];
export const ChgCurrentLimit: [number, string] = [70, 'u16'];
export const ChargeContPwrLmt: [number, string] = [72, 'u16'];
export const ChargePowerLimitExt: [number, string] = [74, 'u16'];
export const ChgExtGpCS: [number, string] = [76, 'u8'];
export const ChgExtGpCounter: [number, string] = [77, 'u8'];
export const ChargeVoltageLimit: [number, string] = [78, 'u16'];
export const CurrentWarning: [number, string] = [80, 'u8'];
export const TempWarning: [number, string] = [81, 'u8'];
export const TempUpLimit: [number, string] = [82, 'i8'];
export const CellVoltWarning: [number, string] = [83, 'u8'];
export const CCCVChargeMode: [number, string] = [84, 'b'];
export const CellVoltUpLimit: [number, string] = [86, 'u16'];
export const SOCHighestCell: [number, string] = [88, 'u16'];
export const SOCLowestCell: [number, string] = [90, 'u16'];
export const SOCAverage: [number, string] = [92, 'u16'];
export const WakeUpTopUpReq: [number, string] = [94, 'b'];
export const WakeUpThermalReq: [number, string] = [95, 'b'];
export const WakeUpDchReq: [number, string] = [96, 'b'];
export const StateofHealth: [number, string] = [98, 'u16'];
export const EstimatedLossChg: [number, string] = [100, 'u16'];
export const CoolingRequest: [number, string] = [102, 'b'];
export const EstimatedLossDch: [number, string] = [104, 'u16'];
export const FanDutyRequest: [number, string] = [106, 'u8'];
export const ValveCtrlStat: [number, string] = [107, 'b'];
export const EstLossDchTgtSoC: [number, string] = [108, 'u16'];
export const HeatPowerGenChg: [number, string] = [110, 'u8'];
export const HeatPowerGenDch: [number, string] = [111, 'u8'];
export const WarmupRateChg: [number, string] = [112, 'u8'];
export const WarmupRateDch: [number, string] = [113, 'u8'];
export const CellVoltageMax: [number, string] = [114, 'u16'];
export const CellVoltageMin: [number, string] = [116, 'u16'];
export const CellTempAverage: [number, string] = [118, 'i8'];
export const CellTempColdest: [number, string] = [119, 'i8'];
export const CellTempHottest: [number, string] = [120, 'i8'];
export const HeaterCtrlStat: [number, string] = [121, 'u8'];
export const ThermalOvercheck: [number, string] = [122, 'b'];
export const InletCoolantTemp: [number, string] = [123, 'i8'];
export const ClntPumpDiagStat_UB: [number, string] = [124, 'b'];
export const InletCoolantTemp_UB: [number, string] = [125, 'b'];
export const CoolantLevel: [number, string] = [126, 'b'];
export const ClntPumpDiagStat: [number, string] = [127, 'b'];
export const MILRequest: [number, string] = [128, 'u8'];
export const EnergyAvailable: [number, string] = [130, 'u16'];
export const EnergyUsableMax: [number, string] = [132, 'u16'];
export const EnergyUsableMin: [number, string] = [134, 'u16'];
export const TotalCapacity: [number, string] = [136, 'u16'];
export const RANGE_ROVER_18B: [number, string] = [144, 'CAN_frame'];
